RecoTracker/LSTCore/interface/alpaka/Constants.h

```diff
@@ -0,0 +1,183 @@
+  template <typename T, typename TAcc, typename TSize>
+  ALPAKA_FN_HOST ALPAKA_FN_INLINE Buf<alpaka::Dev<TAcc>, T> allocBufWrapper(TAcc const& devAccIn, TSize nElements) {
+    return alpaka::allocBuf<T, Idx>(devAccIn, Vec1d(static_cast<Idx>(nElements)));
+  }
+
+  // Wrapper function to reduce code boilerplate for defining grid/block sizes.
+  ALPAKA_FN_HOST ALPAKA_FN_INLINE Vec createVec(int x, int y, int z) {
+    return Vec(static_cast<Idx>(x), static_cast<Idx>(y), static_cast<Idx>(z));
+  }
+
+  // Adjust grid and block sizes based on backend configuration
+  template <typename Vec>
+  ALPAKA_FN_HOST ALPAKA_FN_INLINE WorkDiv createWorkDiv(const Vec& blocksPerGrid,
+                                                        const Vec& threadsPerBlock,
+                                                        const Vec& elementsPerThreadArg) {
+    Vec adjustedBlocks = blocksPerGrid;
+    Vec adjustedThreads = threadsPerBlock;
+
+    // Serial execution, so all launch parameters set to 1.
+#if defined(ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED)
+    adjustedBlocks = Vec::all(static_cast<Idx>(1));
+    adjustedThreads = Vec::all(static_cast<Idx>(1));
+#endif
+
+    // Threads enabled, set number of blocks to 1.
+#if defined(ALPAKA_ACC_CPU_B_SEQ_T_THREADS_ENABLED)
+    adjustedBlocks = Vec::all(static_cast<Idx>(1));
+#endif
+
+    return WorkDiv(adjustedBlocks, adjustedThreads, elementsPerThreadArg);
+  }
+
+// If a compile time flag does not define PT_CUT, default to 0.8 (GeV)
+#ifndef PT_CUT
+  constexpr float PT_CUT = 0.8f;
+#endif
+
+  constexpr unsigned int MAX_BLOCKS = 80;
+  constexpr unsigned int MAX_CONNECTED_MODULES = 40;
+
+  constexpr unsigned int N_MAX_PIXEL_SEGMENTS_PER_MODULE = 50000;
+
+  constexpr unsigned int N_MAX_PIXEL_MD_PER_MODULES = 2 * N_MAX_PIXEL_SEGMENTS_PER_MODULE;
+
+  constexpr unsigned int N_MAX_PIXEL_TRIPLETS = 5000;
+  constexpr unsigned int N_MAX_PIXEL_QUINTUPLETS = 15000;
+
+  constexpr unsigned int N_MAX_PIXEL_TRACK_CANDIDATES = 30000;
+  constexpr unsigned int N_MAX_NONPIXEL_TRACK_CANDIDATES = 1000;
+
+  constexpr unsigned int size_superbins = 45000;
+
+  //defining the constant host device variables right up here
+  // Currently pixel tracks treated as LSs with 2 double layers (IT layers 1+2 and 3+4) and 4 hits. To be potentially handled better in the future.
+  struct Params_pLS {
+    static constexpr int kLayers = 2, kHits = 4;
+  };
+  struct Params_LS {
+    static constexpr int kLayers = 2, kHits = 4;
+  };
+  struct Params_T3 {
+    static constexpr int kLayers = 3, kHits = 6;
+  };
+  struct Params_pT3 {
+    static constexpr int kLayers = 5, kHits = 10;
+  };
+  struct Params_T5 {
+    static constexpr int kLayers = 5, kHits = 10;
+  };
+  struct Params_pT5 {
+    static constexpr int kLayers = 7, kHits = 14;
+  };
+
+  // 15 MeV constant from the approximate Bethe-Bloch formula
+  ALPAKA_STATIC_ACC_MEM_GLOBAL constexpr float kMulsInGeV = 0.015;
```

Comment:
It may be needed for the arrays, if they are used as a whole (i.e. by pointer) instead of only accessing their elements.

Reply:
It's not only a matter of arrays. Transferring here our internal discussion to explain.

---

Let's take as an example the `sinAlphaMax` variable a few lines below. When we remove `ALPAKA_STATIC_ACC_MEM_GLOBAL`  from in front of it, we get:
```
src/RecoTracker/LSTCore/src/alpaka/MiniDoublet.h(352): error: identifier "lst::sinAlphaMax" is undefined in device code
```

This line reads:
`const float miniSlope = alpaka::math::asin(acc, alpaka::math::min(acc, rt * k2Rinv1GeVf / ptCut, sinAlphaMax));`

However, when we change it to:
`const float miniSlope = alpaka::math::asin(acc, alpaka::math::min(acc, rt * k2Rinv1GeVf / ptCut, 1.f * sinAlphaMax));`
it compiles fine.

The function expects a reference (`T const&`) in the arguments. In the example above with `1.f*` the compiler generates a temporary first, which works. But when we use it directly, the `constexpr` (without static specifiers) has no address, hence the issue.

This leads to the following situation:
- For the constants that are always used in conjugation with other constants/numbers, `ALPAKA_STATIC_ACC_MEM_GLOBAL` can be removed.
- For the constants that are used at least once "alone", `ALPAKA_STATIC_ACC_MEM_GLOBAL` is needed.

---

As a result, there can be (at least) two ways forward:
1. We can remove `ALPAKA_STATIC_ACC_MEM_GLOBAL` only for the cases that is not needed, while leaving in it for the others.
2. We can remove `ALPAKA_STATIC_ACC_MEM_GLOBAL` everywhere and create temporaries for the constants that are used "alone".

Let us know what you think would be the best solution (maybe there is another one we haven't thought of).

Reply:
Thank you for the explanation, it's very clear and to the point.

I would use `ALPAKA_STATIC_ACC_MEM_GLOBAL` everywhere.

Reply:
By the way, in the next version of alpaka, the syntax of `ALPAKA_STATIC_ACC_MEM_GLOBAL` is going to change, to address a couple of issues:
  - variable will be templated on the accelerator type, as in `variable<Acc>`; this is to avoid ODR violations and make sure the host and device variables are clearly different;
  - some operations may need to use `variable<Acc>.get()` instead of just `variable<Acc>`; this is to be able to support the new SYCL/oneAPI back-ends.

Anyway, this is just a heads up, the time scale for these changes is very likely post-CMSSW 14.1.0.

Reply:
Maybe add a short comment that explains the reason for the use of `ALPAKA_STATIC_ACC_MEM_GLOBAL` in addition of `constexpr` (mentioning `alpaka::math::min()` / `std::min()` argument types would be sufficient for me).